src/v/security/gssapi_authenticator.cc

```diff
@@ -111,55 +112,126 @@ void display_status(
     display_status_1(msg, min_stat, GSS_C_MECH_CODE);
 }
 
+// impl runs on the thread_worker
+class gssapi_authenticator::impl {
+public:
+    using state = gssapi_authenticator::state;
+
+    template<typename R>
+    struct state_result {
+        gssapi_authenticator::state state;
+        result<R> result;
+    };
+
+    impl(
+      ss::sstring primary, ss::sstring keytab, std::vector<gssapi_rule> rules)
+      : _primary{std::move(primary)}
+      , _keytab{std::move(keytab)}
+      , _rules{std::move(rules)} {}
+
+    state_result<bytes> authenticate(bytes auth_bytes);
+    const security::acl_principal& principal() const { return _principal; }
+
+private:
+    state_result<void> init();
+    state_result<bytes> more(bytes_view);
+    state_result<bytes> ssfcap(bytes_view);
+    state_result<bytes> ssfreq(bytes_view);
+    state_result<void> check();
+    void
+    fail_impl(OM_uint32 maj_stat, OM_uint32 min_stat, std::string_view msg);
+    template<typename... Args>
+    void fail(
+      OM_uint32 maj_stat,
+      OM_uint32 min_stat,
+      fmt::format_string<Args...> format_str,
+      Args&&... args) {
+        auto msg = ssx::sformat(format_str, std::forward<Args>(args)...);
+        fail_impl(maj_stat, min_stat, msg);
+    }
+    acl_principal get_principal_from_name(std::string_view source_name);
+
+    ss::sstring _primary;
+    ss::sstring _keytab;
+    const std::vector<gssapi_rule> _rules;
+    security::acl_principal _principal;
+    state _state{state::init};
+    gss::cred_id _server_creds;
+    gss::ctx_id _context;
+};
+
+gssapi_authenticator::gssapi_authenticator(
+  ssx::thread_worker& thread_worker, std::vector<gssapi_rule> rules)
+  : _worker{thread_worker}
+  , _impl{std::make_unique<impl>(
+      config::shard_local_cfg().sasl_kerberos_principal(),
+      config::shard_local_cfg().sasl_kerberos_keytab(),
+      std::move(rules))} {}
+
+gssapi_authenticator::~gssapi_authenticator() = default;
+
 ss::future<result<bytes>> gssapi_authenticator::authenticate(bytes auth_bytes) {
+    if (!_impl) {
+        vlog(
+          seclog.warn,
+          "authenticate received after handshake complete {} bytes",
+          _state,
+          auth_bytes.size());
+        co_return errc::invalid_gssapi_state;
```

Comment:
Is there a real possibility of this happening?  Would this be a candidate for a `vassert`?

Reply:
> Is there a real possibility of this happening?  Would this be a candidate for a `vassert`?

The connection is open to the "public", so it avoids a type of DoS.

I tend to use vassert only if the logic is wrong, rather than bad user input

Reply:
No I understand what you mean.  My question was more geared towards would the code itself be wrong if we ended up in this situation.

Reply:
More of ignorance on my part (or lack of sleep to grasp a simple situation) but I'm trying to think how an instance of `authenticator` get reused.

Reply:
I guess so, but the transition to `complete` is pretty simple at the moment. Any transition to `failed` should tear down the connection anyway.

Reply:
> More of ignorance on my part (or lack of sleep to grasp a simple situation) but I'm trying to think how an instance of `authenticator` get reused.

It's tied to the connection